### PR TITLE
FM-174

### PR DESCRIPTION
**Pull request by milankarunarathne**
@harsha89  @surangak please, review this merge request.

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/fhir/server/FHIRRESTServer.java [hunk 1]
```diff
@@ -74,7 +74,7 @@ protected void initialize() throws ServletException {
 		setDefaultPrettyPrint(true);
 		setDefaultResponseEncoding(EncodingEnum.JSON);
 		if (FHIRUtils.isCustomNarrativesEnabled()) {
-			String propFile = FHIRUtils.gettCustomNarrativesPropertyPath();
+			String propFile = FHIRUtils.getCustomNarrativesPropertyPath();
```
Comment (surangak) [inline omod/src/main/java/org/openmrs/module/fhir/server/FHIRRESTServer.java]: Good catch :-)

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java [hunk 2]
```diff
@@ -24,15 +64,314 @@ public String getServiceCategory() {
...
+			//TODO: org.openmrs.Patient omrsParient = FHIRPatientUtil.generateOpenMRSPatient(patient);
+			omrsPatient = new org.openmrs.Patient();
+			omrsDiagnosticReport.setPatient(omrsPatient);
+		} else {
+			// Get Id of the Patient
+			String patientID = diagnosticReport.getSubject().getReference().getIdPart();
+			// Retrieve Patient
+			String serverBase = FHIRUtils.getDiagnosticReportRadiologyBaseServerURL();
+			Patient patient = FHIRRESTfulGenericClient.readPatientById(serverBase, patientID);
+			System.out.println("Patient " + patient.getId().getIdPart());
```
Comment (surangak) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java]: Death to sysouts :-) switch to logging

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/DefaultDiagnosticReportHandler.java [hunk 3]
```diff
@@ -36,6 +43,11 @@ public DiagnosticReport getFHIRDiagnosticReportById(String id) {
 	}
 
 	@Override
+	public List<DiagnosticReport> getFHIRDiagnosticReportBySubjectName(String name) {
+		return new ArrayList<DiagnosticReport>();
+	}
+
+	@Override
 	public DiagnosticReport saveFHIRDiagnosticReport(DiagnosticReport diagnosticReport) {
 		System.out.println("Laboratory Handler : diSave FHIR Diagnostic Report");
```
Comment (surangak) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/DefaultDiagnosticReportHandler.java]: Death to sysouts :)

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java [hunk 4]
```diff
@@ -24,15 +64,314 @@ public String getServiceCategory() {
...
+	/**
+	 * Getting the data from a third party server, storing metadata in OpenMRS and passing on the bundle to the requester.
+	 *
+	 * @param name Given Name of the Subject/Patient
+	 * @return A bundle of DiagnosticReport which is matching with Given Name
+	 */
+	@Override
+	public List<DiagnosticReport> getFHIRDiagnosticReportBySubjectName(String name) {
+		log.debug("In Radiology Handler : getFHIRDiagnosticReportBySubjectName : " + name);
+		ArrayList<DiagnosticReport> diagnosticReports = new ArrayList<DiagnosticReport>();
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java]: Use abstract classes like List<DiagnosticReport> diagnosticReports = new ArrayList<DiagnosticReport>();

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java [hunk 5]
```diff
@@ -24,15 +64,314 @@ public String getServiceCategory() {
...
+
+		String serverBase = FHIRUtils.getDiagnosticReportRadiologyBaseServerURL();
+		ICriterion<ReferenceClientParam> diagnosticReportBySubject = DiagnosticReport.SUBJECT.hasChainedProperty(
+				Patient.GIVEN.matches().value(name));
+		ICriterion<TokenClientParam> diagnosticReportByService = DiagnosticReport.SERVICE.exactly().code("RAD");
+		Bundle bundle = FHIRRESTfulGenericClient.search(serverBase, DiagnosticReport.class, diagnosticReportBySubject,
+				diagnosticReportByService);
+		log.debug("Bundle size : " + bundle.size());
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java]: normally use if(log.isDebugEnable()) {
log.debug(msg)
}
tHis is to improve performance

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java [hunk 6]
```diff
@@ -24,15 +64,314 @@ public String getServiceCategory() {
...
+
+		for (BundleEntry entry : bundle.getEntries()) {
+			IResource resource = entry.getResource();
+			log.debug("Resource Type : " + resource.getResourceName());
+			if (resource.getResourceName().equals("DiagnosticReport")) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java]: normally use if(log.isDebugEnable()) {
log.debug(msg)
}
tHis is to improve performance

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java [hunk 7]
```diff
@@ -24,15 +64,314 @@ public String getServiceCategory() {
...
+				diagnosticReport = this.getFHIRDiagnosticReportById(diagnosticReport.getId().getIdPart());*/
+				diagnosticReports.add(diagnosticReport);
+			}
+		}
+
+		return diagnosticReports;
+	}
+
+	private DiagnosticReport getFHIRDiagnosticReport(Encounter omrsDiagnosticReport) {
+		log.debug("Laboratory Handler : GetFHIRDiagnosticReport");
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java]: normally use if(log.isDebugEnable()) {
log.debug(msg)
}
tHis is to improve performance

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java (hunk 2 above)
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java]: souts :(

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java [hunk 8]
```diff
@@ -24,15 +64,314 @@ public String getServiceCategory() {
...
+			org.openmrs.Provider omrsProvider = FHIRPractitionerUtil.generatePractitioner();*/
+			Provider omrsProvider = new Provider();
+			omrsDiagnosticReport.setProvider(new EncounterRole(), omrsProvider);
+		} else {
+			// Get Id of the Performer
+			String practitionerID = diagnosticReport.getPerformer().getReference().getIdPart();
+			// Retrieve Practitioner
+			String serverBase = FHIRUtils.getDiagnosticReportRadiologyBaseServerURL();
+			Practitioner practitioner = FHIRRESTfulGenericClient.readPractitionerById(serverBase, practitionerID);
+			System.out.println("Practitioner " + practitioner.getId().getIdPart());
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/RadiologyHandler.java]: souts :(

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+public class FHIRRESTfulGenericClient {
+
+	private static final FhirContext ctx = FhirContext.forDstu2();
+
+	public static Bundle searchDiagnosticReportByReference(String serverBase,
+	                            ICriterion<ReferenceClientParam> where) {
+		IGenericClient client = ctx.newRestfulGenericClient(serverBase);
+
+		return client
+				.search()
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java]: some formatting issues

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	public static Patient readPatientById(String serverBase,
+	                                     String theID) {
+		IGenericClient client = ctx.newRestfulGenericClient(serverBase);
+
+		return client
+				.read()
+				.resource(Patient.class)
+				.withId(theID)
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java]: some formatting issues

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+				.execute();
+	}
+
+	public static Practitioner readPractitionerById(String serverBase,
+	                                      String theID) {
+		IGenericClient client = ctx.newRestfulGenericClient(serverBase);
+
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java]: some formatting issues

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+		return client
+				.read()
+				.resource(Practitioner.class)
+				.withId(theID)
+				.execute();
+	}
+
+	public static Bundle search(String serverBase,
+	                            Class<DiagnosticReport> fhirResource,
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java]: some formatting issues

**Comment by surangak** [comment]
Yay! address these comments and we're ready to merge!!! :)

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/DefaultDiagnosticReportHandler.java (hunk 3 above)
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/DefaultDiagnosticReportHandler.java]: @milankarunarathne seems u missed this?

File: pom.xml [hunk 13]
```diff
@@ -33,6 +33,10 @@
             <name>harsz89</name>
             <email>harsz89@gmail.com</email>
         </developer>
+        <developer>
+            <name>milankarunarathne</name>
+            <email>mhkarunarathne@gmail.com</email>
+        </developer>
     </developers>
```
Comment (surangak) [inline pom.xml]: @sashrika , your next commit should add your creds here :)

Comment (sashrika) [inline pom.xml]: @surangak  I will :)

**Pull request merged**